bentoml/_internal/bento/gen.py

```diff
@@ -84,28 +83,89 @@ class CustomizableEnv:
     path: str = attr.field(default=BENTO_PATH)
 
 
-bentoml_cattr.register_unstructure_hook(
+cattr.register_unstructure_hook(
     ReservedEnv, lambda rs: {f"__{k}__": v for k, v in attr.asdict(rs).items()}
 )
 attr.resolve_types(ReservedEnv, globals(), locals())
 
-bentoml_cattr.register_unstructure_hook(
+cattr.register_unstructure_hook(
     CustomizableEnv, lambda rs: {f"bento__{k}": v for k, v in attr.asdict(rs).items()}
 )
 
 attr.resolve_types(CustomizableEnv, globals(), locals())
 
+TEMPLATES_PATH = [fs.path.join(os.path.dirname(__file__), "docker", "templates")]
+ENVIRONMENT = Environment(
+    extensions=["jinja2.ext.do", "jinja2.ext.loopcontrols", "jinja2.ext.debug"],
+    trim_blocks=True,
+    lstrip_blocks=True,
+    loader=FileSystemLoader(TEMPLATES_PATH, followlinks=True),
+)
+
 
-def get_template_env(
-    docker_options: DockerOptions, spec: DistroSpec
+def validate_user_template(template: Template, loader: BaseLoader) -> None | t.NoReturn:
+    """
+    Validate all user-defined templates in the given environment.
+    """
+    ctx = template.new_context()
+    if "bento__dockerfile" not in ctx.keys():
+        exc_info = (
+            f"User-defined template `{template}` does not contain `bento__dockerfile`. "
```

Comment:
why is that a problem?  user should be able to generate everything from scratch without our base templates right?

Reply:
the user could also be just extending one block? right?

Reply:
actually we need they to have our base template to make sure that bento still works